### ISSUE TITLE
Add a new automatic variable that uniquely identifies a single test case run

### ISSUE THREAD

**Issue by Ionsilver** [robotframework/robotframework#2743]
## Description

I suggest adding a new automatic variable, named for example `${TEST RUN ID}`, available in test cases, that would uniquely identify a single run of the test case. 

This means that:
* In a single run, each test case (and in the case of a template test case, each instance of the template) would have a different value of ${TEST RUN ID}.
* In two different runs (for example, in two runs started with RIDE or the maven plugin), the same test case would have different values of ${TEST RUN ID} in each run.

I propose to do this by setting the value of ${TEST RUN ID} to "TIMESTAMP-ID" where TIMESTAMP is the Unix timestamp of the time the entire test run was started and ID is the zero-based index of the test case in an execution context. This value and the variable name are just suggestions, all that this use case needs is the the variable value is reasonably unique.

I would be happy to provide a pull request.

## Use case

Test cases may affect the system under test. Some test cases create new entities in the system under test that must be identified somehow so the test case may refer to them afterwards. ${TEST RUN ID} would be used, for example, as part of a new user's username. 

An alternative would be to use fixed strings for this purpose, but that would require teardowns to not fail, or even be available. In the previous use case, the system under test might not support deleting users.

Another alternative would be to generate this string in user code, for example in Test Setup. This is what I currently do, but this variable is useful in almost all tests, and Test Setup may be overriden by tests so it's not sufficient to declare it only in the root suite Test Setup. It also not possible to put this code in the root suite Suite Setup because the variable value needs to change for each test.

## Backward compatibility concerns

Because automatic variables have higher priority than everything except setting variables from keywords, test cases which depend on a specific value of a variable named ${TEST RUN ID} being set in a variable table, by command line or in a variable file will fail.

**Comment by bkhouri** [comment]
An alternative to using fixed string is to use randomly generated strings, that's personally what I do. Just create a custom "random generator", store a random string in a test suite, and use that everywhere.   In the use case of creating a username, I would have something like this:

```robotframework
*** Settings ***
Library    MyCustomRandomGeneratorLibrary     WITH NAME  RandomGenerator
Suite Setup    Create Suite Variables

*** Keywords ***
Create Suite Variables
     ${username}=     RandomGenerator.random_alphanumeric
     ${password}=     RandomGenerator.random_alphanumeric
     ${email}=        RandomGenerator.email_address
     Set Suite Variable  ${username}
     Set Suite Variable  ${password}
     Set Suite Variable  ${email}

*** Test Cases ***
Add user
     ${userinfo}=     Add User      ${username}   ${password}  ${email} 
     Verify UserName  ${userinfo}   ${username}     # need to define this keywords
     Verify Password  ${userinfo}   ${password}     # need to define this keywords
     Verify email     ${userinfo}   ${email}        # need to define this keywords
```

**Comment by Ionsilver** [comment]
@bkhouri That is similar to what I do as well. I would like to avoid having to generate the string in Test Setup, because I need it almost everywhere.

**Comment by laurentbristiel** [comment]
Interesting idea. I think I could use it and I guess many of us have this need of a unique "id" to create object/entity within a test (and that would not have risk of collision with other objects/entities in the system). Did you consider UUID instead of timestamp? That would be a bit more random-like.

**Comment by Ionsilver** [comment]
A UUID is another option. It would have the advantage of being unique even if two tests are run at the exact same millisecond (which might happen if tests are run in parallel). It would have the disadvantage that it's long (36 characters) and contains special characters, so it might not be permissible in all places (such as username in the example).

**Comment by pekkaklarck** [comment]
Some comments below:

1. Having a unique identifier sounds useful, but I'm not sure can we come up with an identifier that is suitable in different use cases. As already discussed, timestamps would work in some cases but in others UUIDs might be better, and even then we needed to decide the detailed format.

2. My recommendation to create an identifier for the test run nowadays would be doing that in a start-up script and passing it in using the `--variable` option. Alternatively you could have a variable file that creates the variable that you give with the `--variablefile` option. The benefit of this approach is that you can use timestamps, UUIDs, or whatever suites your needs. If we cannot come up with a generic identifier suitable in most use cases, perhaps this is enough.

3. Did I get it right that the identifier would change during the execution round depending on which test is currently executing? I would consider that a bit strange, because I'd assume you wanted to be able to refer to the *same* unique identifier in different tests. Tests having an id would be useful, though, and Robot actually already internally assigns ids to tests, suites and keywords in format like `s1.t2` (second test in first suite). Perhaps we should expose these ids as variables? If there were a unique identifier for the whole run, either provided by the framework or set when the execution starts, then they could be used in combination like `${TEST RUN ID}-${TEST ID}` if needed.

**Comment by Ionsilver** [comment]
A value such as ${RUN ID}-${TEST ID} would be what I had in mind, yes (where RUN ID is an identifier common to all the test executions in a single mvn command/RIDE test runner launch).

The value needs to change depending on which test is currently executing, because I want the tests to be independent. If test 1 creates a user with name "username-${TEST RUN ID}", and then fails to delete it, I want to both be able to use "username-${TEST RUN ID}" in the same test next execution (which I could if it didn't change) and to be able to use the same name in another test (which I couldn't). This way, the tests won't interfere with each other.

**Comment by pekkaklarck** [comment]
Adding an UUID seems to be a common need (see the newer #2809) and feels like useful addition in general for me. After thinking this a bit more, I propose this kind of design:

1. Generate a unique ID for each run as n UUID. We can use Python's [uuid module](https://docs.python.org/2/library/uuid.html) but need to still decide which method.

2. Allow setting the UUID from the command line with, for example, `--id` or `--executionid` option. This is optional and could be implemented later.

3. Add the execution ID to the top level suite as [free metadata](http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#free-test-suite-metadata). This automatically makes it visible in logs and reports, available in output.xml for external tools, and so on. Name for the metadata could be `Execution ID`. Alternatively we could add the execution ID as new attribute to suite objects but that would be a lot bigger task.

4. Expose the execution ID as a global variable. Variable name could be `${EXECUTION ID}`.

5. Expose existing test and suite IDs and possibly also keyword ID as [automatic variables](http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#automatic-variables). These IDs are in format like `s1.s2.t5` and are mainly used internally. Allows creating unique IDs for suites and tests like `${EXECUTION ID}-${TEST ID}`. This requires a separate issue.

6. Decide how to handle this when combining results with Rebot. Most likely the newly created top-level suite should get its own execution ID, but need to think should existing IDs from lower level suites be removed. One thing to take into account is that the aforementioned test and suite IDs are not static and combining results recreates them.

Comments about the above design is more than welcome. More importantly, we'd need someone interested to implement (incl. tests and documentation) all this. This would be a good addition to RF 3.1 scheduled for early June, but I doubt I have time to implement this. I'll obviously review possible PRs and help otherwise. I don't think implementing this would be overly complicated but there's quite a lot of work.

**Comment by Ionsilver** [comment]
I won't be able to make use of this because many of our database fields are limited to 40 characters' length, and a UUID combined with a test id (which is what I would need under this proposal) would exceed that, even without adding any descriptive text. I would instead continue to use `Test Setup` or a listener library.

I don't know how common my scenario is, though.

**Comment by pekkaklarck** [comment]
If we want the ID to be automatically generated and unique, it needs to be pretty long to make the possibility for collisions small. According to the [RFC 4122](https://tools.ietf.org/html/rfc4122.html) that Python's `uuid` module implements UUIDs are 128 bit long which means 32 characters in hex plus possible hyphens used between the blocks. I doubt there is any other scheme to create IDs that would yield values that are considerably shorted while keeping them unique.

I proposed above the possibility to give the ID externally using `--executionid` option to allow using some other ID scheme. We could make that even better and allow user to select which scheme Robot should use automatically. Robot could, for example, have the following ID schemes and allow user to select between them like `--executionid epoch`.

- UUID: Create the ID using Python's `uuid` module. On a closer look at the docs, it seems `uuid.uuid4()` would work best. The result would be something like `7cb71896-2adb-4dcb-b8a9-2e1d44f8ee01` (36 characters). Ought to be always unique.

- Timestamp: Use timestamp in format `data-time-micro` resulting to something like `20180417-174644-359409` (22 characters). This is unique unless tests are run in parallel (or clocks are not in sync) and even then collisions are very rare.

- Epoch: Use epoch timestamp converted to hex to get same characteristics as above but shorted value. Could be created, for example, with `hex(int(time.time()*1000000))[2:]` and result would be something like `56a0c5302c86b` (13 characters).

- Custom: Use a custom value. The actual value should be given as part of the value, for example, like `--executionid custom:1`.
